provider/resource_keycloak_realm.go

```diff
@@ -456,9 +513,47 @@ func getRealmFromData(data *schema.ResourceData) (*keycloak.Realm, error) {
 		realm.ActionTokenGeneratedByAdminLifespan = actionTokenGeneratedByAdminLifespanDurationString
 	}
 
+	//security defenses
+	if v, ok := data.GetOk("security_defenses"); ok {
+		securityDefensesSettings := v.([]interface{})[0].(map[string]interface{})
+
+		headersConfig := securityDefensesSettings["headers"].([]interface{})
+		if len(headersConfig) == 1 {
+			headerSettings := headersConfig[0].(map[string]interface{})
+
+			realm.Attributes = keycloak.Attributes{
+				BrowserHeaderContentSecurityPolicy:           headerSettings["content_security_policy"].(string),
+				BrowserHeaderContentSecurityPolicyReportOnly: headerSettings["content_security_policy_report_only"].(string),
+				BrowserHeaderStrictTransportSecurity:         headerSettings["strict_transport_security"].(string),
+				BrowserHeaderXContentTypeOptions:             headerSettings["x_content_type_options"].(string),
+				BrowserHeaderXFrameOptions:                   headerSettings["x_frame_options"].(string),
+				BrowserHeaderXRobotsTag:                      headerSettings["x_robots_tag"].(string),
+				BrowserHeaderXXSSProtection:                  headerSettings["x_xss_protection"].(string),
+			}
+		} else {
+			setDefaultSecuritySettings(realm)
+		}
+		//todo bruteforcedetection
+	} else {
+		setDefaultSecuritySettings(realm)
+	}
 	return realm, nil
 }
 
+func setDefaultSecuritySettings(realm *keycloak.Realm) {
+	realm.Attributes = keycloak.Attributes{
+		BrowserHeaderContentSecurityPolicy:           "frame-src 'self'; frame-ancestors 'self'; object-src 'none';",
+		BrowserHeaderContentSecurityPolicyReportOnly: "",
+		BrowserHeaderStrictTransportSecurity:         "max-age=31536000; includeSubDomains",
+		BrowserHeaderXContentTypeOptions:             "nosniff",
+		BrowserHeaderXFrameOptions:                   "SAMEORIGIN",
+		BrowserHeaderXRobotsTag:                      "none",
+		BrowserHeaderXXSSProtection:                  "1; mode=block",
+
+		//todo add default bruteforce settings
+	}
+}
```

Comment:
I'm not sure I understand the purpose of this function if the schema is already setting these defaults, could you clear this up for me?

Reply:
The problem is, when you remove the 'headers' block in terraform, the values that you had set before, will persist and will not be reset to the keycloak defaults. 
You can also not make them empty as then terraform will send empty strings(which are valid values) to keycloak and thus keycloak will not work anymore as intended,
So my idea is, if there is no 'headers' block, then set values to sensible defaults. (Imho, these are general sensible defaults and not limited to keycloak)

Furthermore the default values set in scheme, will only be used when there is a 'headers' block. This way, users only need to fill in headers to are different from the default.
When there is no headers block terraform does something like this:

example
1. Set x frame options to DENY 
2. remove security defenses and headers block from main.tf (without the setDefaultSecuritySettings(realm) method)

Terraform will perform the following actions:
```
  ~ keycloak_realm.test
      security_defenses.#:                           "1" => "0"
      security_defenses.0.headers.#:                 "1" => "0"
      security_defenses.0.headers.0.x_frame_options: "DENY" => "SAMEORIGIN"
```

But in Keycloak this will remain on "DENY"

with the setDefaultSecuritySettings(realm) method, this works as intended. also see test TestAccKeycloakRealm_securityDefenses

I agree this feels like a workaround/patch solution, But I did not see any other solution.
If you have a better solution to make this work transparently for users, I am happy to change this implementation.

Reply:
Hmm, I'm torn on this one.  As a user, would you expect that deleting an attribute would cause a change to be made within Keycloak, or for that attribute to simply not be managed by Terraform anymore?

I think I'm in favor of the latter.  Most of my Terraform experience is with the Google provider, and a lot of their resources behave this way - if an optional field is deleted, it doesn't change it back to the perceived default, it simply stops caring about it.

I understand the point you're making, I'm just not sure what kind of behavior a typical Terraform user would expect here.

Reply:
I dont agree on the latter.. If you delete an attribute from terraform it will delete or change that managed resource, it is only when you remove the current state of that resource & delete the attribute from the terraform code that it is "forgotten" or no longer managed.